appmonitoring/ts/src/CertificateGenerator.ts

```diff
@@ -0,0 +1,229 @@
+import * as k8s from "@kubernetes/client-node";
+import forge from "node-forge";
+import pem from "pem";
+
+class WebhookCertData {
+    caCert: string;
+    tlsCert: string;
+    tlsKey: string;
+}
+
+const hostname = process.env.HOSTNAME;
+
+export class CertificateManager {
+
+    private static makeNumberPositive = (hexString) => {
+        let mostSignificativeHexDigitAsInt = parseInt(hexString[0], 16);
+    
+        if (mostSignificativeHexDigitAsInt < 8) return hexString;
+    
+        mostSignificativeHexDigitAsInt -= 8
+        return mostSignificativeHexDigitAsInt.toString() + hexString.substring(1)
+    }
+    
+    // Generate a random serial number for the Certificate
+    private static randomSerialNumber = () => {
+        return CertificateManager.makeNumberPositive(forge.util.bytesToHex(forge.random.getBytesSync(20)));
+    }
+
+    private static async GenerateSelfSignedCertificate(): Promise<WebhookCertData> {
+
+        let caCert: forge.pki.Certificate = forge.pki.createCertificate();
+        let keys = forge.pki.rsa.generateKeyPair(4096);
+        caCert.serialNumber = CertificateManager.randomSerialNumber();
+        caCert.publicKey = keys.publicKey;
+        caCert.privateKey = keys.privateKey;
+        caCert.validity.notBefore = new Date(2023,6,20,0,0,0,0);
+        caCert.validity.notAfter = new Date(2025,7,21,0,0,0,0);
+
+        const attributes = [{
+            shortName: 'CN',
+            value: 'applicationinsights-ca'
+        }];
+        caCert.setSubject(attributes);
+        caCert.setIssuer(attributes);
+    
+        const extensions = [{
+            name: 'basicConstraints',
+            cA: true
+        },{
+            name: 'subjectKeyIdentifier',
+            keyIdentifier: caCert.generateSubjectKeyIdentifier().getBytes(),
+        },{
+            name: 'keyUsage',
+            keyCertSign: true,
+            cRLSign: true,
+            digitalSignature: true,
+            keyEncipherment: true,
+        }];
+
+        caCert.setExtensions(extensions);
+        caCert.sign(caCert.privateKey,forge.md.sha256.create());
+
+        const caCertResult: pem.CertificateCreationResult = {
+            certificate: forge.pki.certificateToPem(caCert),
+            serviceKey: forge.pki.privateKeyToPem(caCert.privateKey),
+            csr: undefined,
+            clientKey: undefined
+            
+        }
+
+        const host_attributes = [{
+            shortName: 'CN',
+            value: "app-monitoring-webhook-service.kube-system.svc"
```

Comment:
app-monitoring-webhook-service.kube-system.svc has to be in one place and used in the following places:
- here inside of the image to generate the cert
- in yaml where kind: Service is generated (that is a front gate for webhook pods)
- in MutatingWebHookConfiguration where the webhook endpoint is specified

Reply:
Can we have a single source of truth for this value app-monitoring-webhook-service.kube-system.svc? Right not it's here and also in AKS RP yaml, can't change one place without changing the other, very error-prone.

Reply:
Unfortunately not, as there is no way to synchronize the two, the best we can do is write unit tests that validate this name. That way you cannot accidentally change it?